src/main/scala/cloud/drdrdr/utils/AWSCredentials.scala

```diff
@@ -6,165 +6,416 @@ package cloud.drdrdr.utils
+    def getCredentials: Future[AWSPermissions] = {
+      credentials flatMap {
+        case perm:AWSPermissions =>
+          if (perm.expiration.isEmpty) {
+              credentials
+          } else {
+              checkExpire flatMap {
+                case bol =>
+                  if (bol) {
+                    credentials = updateCredentials()
+                  }
+                  credentials
+              }
+          }
+      }
+    }
+
+    //checks if current credentials have expired
+    private def checkExpire()(implicit ec: ExecutionContext, s: ActorSystem, m: ActorMaterializer): Future[Boolean] = {
+      credentials map {
+        case creds: AWSPermissions =>
+          if (creds.expiration.isEmpty)
+            false
+          else
+            creds.expiration <= getUTCTime()
+      }
+    }
+//
+    //updates the credentials on an EC2 instance if necessary
+    @throws(classOf[Exception])
+    private def updateCredentials(): Future[AWSPermissions] = {
+      getAmazonEC2Credentials() map {
+        case Some(cred) => cred
+        case None => throw new Exception("Unable to update credentials")
+      }
+    }
+  }
 
+  /**
+   *
+   * @param keyId aws accessKeyId
+   * @param secretKey aws accessSecretKey
+   * @param tok aws token associated with credentials
+   * @param exp aws expiration of credentials
+   */
+  class AWSPermissions(keyId: String, secretKey: String, tok: String = "", exp: String = "") {
+    private val accessKey = keyId
+    private val secretAccess = secretKey
+    private val t = tok
+    private val e = exp
 
-object AWSCredentials {
+    /**
+     * accessKeyId accessor method
+     * @return accessKeyId
+     */
+    def accessKeyId: String = {
+      this.accessKey
+    }
 
-  implicit val system: ActorSystem = ActorSystem()
-  implicit val ec: ExecutionContext = system.dispatcher
-  implicit val materializer: ActorMaterializer = ActorMaterializer()
+    /**
+     * secretAccessKey accessor method
+     * @return secretAccessKey
+     */
+    def secretAccessKey: String = {
+      this.secretAccess
+    }
 
-  case class AWSPermissions(accessKeyId: String, secretAccessKey: String, token: String = "")
+    /**
+     * token accessor method
+     * @return token
+     */
+    def token: String = {
+      this.t
+    }
+
+    /**
+     * expiration accessor method
+     * @return expiration
+     */
+    def expiration: String = {
+      this.e
+    }
+  }
 
-  def validCredentials(key_id:Option[String], access_key:Option[String], token: Option[String] = Some("")): Option[AWSPermissions] = {
-    if (key_id.isDefined && access_key.isDefined && key_id.get != null && access_key.get != null)
-      Some(AWSPermissions(key_id.get, access_key.get, token.get))
+  //checks if both the access key Id and the secret key are valid
+  def validCredentials(keyId: Option[String], accessKey: Option[String], token: Option[String] = Some(""), expiration: Option[String] = Some("")): Option[AWSPermissions] = {
```

Comment:
I would still have to give them default values. I guess now those could be None to make look a little cleaner.